src/renderers/WebGLRenderer.js

```diff
@@ -861,6 +861,162 @@ function WebGLRenderer( parameters = {} ) {
 
 	};
 
+	// compileAsync
+
+	this.compileAsync = function ( scene, targetScene = null ) {
+
+		// If no explicit targetScene was given use the scene instead
+		if ( ! targetScene ) {
+
+			targetScene = scene;
+
+		}
+
+		currentRenderState = renderStates.get( targetScene );
+		currentRenderState.init();
+
+		renderStateStack.push( currentRenderState );
+
+		let foundScene = scene === targetScene;
+
+		// Gather lights from both the scene and the new object that will be added
+		// to the scene.
+		targetScene.traverseVisible( function ( object ) {
+
+			if ( object === scene ) {
+
+				foundScene = true;
+
+			}
+
+			if ( object.isLight ) {
+
+				currentRenderState.pushLight( object );
+
+				if ( object.castShadow ) {
+
+					currentRenderState.pushShadow( object );
+
+				}
+
+			}
+
+		} );
+
+		// If the scene wasn't already part of the targetScene, add any lights it
+		// contains as well.
+		if ( ! foundScene ) {
+
+			scene.traverseVisible( function ( object ) {
+
+				if ( object.isLight ) {
+
+					currentRenderState.pushLight( object );
+
+					if ( object.castShadow ) {
+
+						currentRenderState.pushShadow( object );
+
+					}
+
+				}
+
+			} );
+
+		}
+
+		currentRenderState.setupLights( _this.physicallyCorrectLights );
+
+		const compiling = new Set();
+
+		// Only initialize materials in the new scene, not the targetScene.
+
+		scene.traverse( function ( object ) {
+
+			const material = object.material;
+
+			if ( material ) {
+
+				if ( Array.isArray( material ) ) {
+
+					for ( let i = 0; i < material.length; i ++ ) {
+
+						const material2 = material[ i ];
+
+						getProgram( material2, targetScene, object );
+						compiling.add( material2 );
+
+					}
+
+				} else {
+
+					getProgram( material, targetScene, object );
+					compiling.add( material );
+
+				}
+
+			}
+
+		} );
+
+		currentRenderState = null;
+
+		// Wait for all the materials in the new object to indicate that they're
+		// ready to be used before resolving the promise.
+
+		return new Promise( ( resolve ) => {
```

Comment:
Kind of minor but... could we use async/await instead?

Reply:
I don't think it is in this case. The problem is that this kicks off a polling loop with `setTimeout()`, which of course is callback based. If we want to treat it as a promise we have to explicitly wrap it in a `Promise()` constructor.

This does still allow developers to call the function as `await renderer.compileAsync(....);` though! And I've updated the two examples this was used in to demonstrate exactly that, because it's a much nicer pattern.